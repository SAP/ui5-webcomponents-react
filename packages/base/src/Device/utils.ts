import { getOS, supportTouch } from '@ui5/webcomponents-base/dist/Device.js';
import { supportMatchMedia, supportOrientation } from './Support';

export const getWindowSize = () => {
  return [window.innerWidth, window.innerHeight];
};

export const changeRootCSSClass = (sClassName, bRemove, sPrefix?) => {
  const oRoot = document.documentElement;
  if (oRoot.className.length === 0) {
    if (!bRemove) {
      oRoot.className = sClassName;
    }
  } else {
    const aCurrentClasses = oRoot.className.split(' ');
    let sNewClasses = '';
    aCurrentClasses.forEach((currentClass) => {
      // eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
      if ((sPrefix && currentClass.indexOf(sPrefix) !== 0) || (!sPrefix && currentClass !== sClassName)) {
        sNewClasses = `${sNewClasses + currentClass} `;
      }
    });
    if (!bRemove) {
      sNewClasses += sClassName;
    }
    oRoot.className = sNewClasses;
  }
};

export const getQuery = (from: number, to: number, unit = 'px') => {
  let q = 'all';
  if (from > 0) {
    q = `${q} and (min-width:${from}${unit})`;
  }
  if (to > 0) {
    q = `${q} and (max-width:${to}${unit})`;
  }
  return q;
};

export const convertToPx = (val, unit?) => {
  if (unit === 'em' || unit === 'rem') {
    const s = window.getComputedStyle ?? ((element: any) => element.currentStyle);
    const x = s(document.documentElement).fontSize;
    const f = x && x.indexOf('px') >= 0 ? parseFloat(x) : 16;
    return val * f;
  }
  return val;
};

export const matchLegacyBySize = (from, to, unit, size) => {
  from = convertToPx(from, unit);
  to = convertToPx(to, unit);

  const width = size[0];
  const a = from < 0 || from <= width;
  const b = to < 0 || width <= to;
  return a && b;
};

export const isLandscape = (bFromOrientationChange: boolean, orientation, bKeyboardOpen: boolean) => {
  if (supportTouch() && supportOrientation() && getOS().android) {
    // if on screen keyboard is open and the call of this method is from orientation change listener,
    // reverse the last value. this is because when keyboard opens on android device, the height can be less
    // than the width even in portrait mode.
    if (bKeyboardOpen && bFromOrientationChange) {
      return !orientation.landscape;
    }
    if (bKeyboardOpen) {
      // when keyboard opens, the last orientation change value will be returned.
      return orientation.landscape;
    }
  } else if (supportMatchMedia() && supportOrientation()) {
    // most desktop browsers and windows phone/tablet which not support orientationchange
    return !!window.matchMedia('(orientation: landscape)').matches;
  }
  // otherwise compare the width and height of window
  const size = getWindowSize();
  return size[0] > size[1];
};
