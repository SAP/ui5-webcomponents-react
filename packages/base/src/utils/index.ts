import { UIEvent } from 'react';

export const deprecationNotice = (component: string, message: string) => {
  if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
    const value = `*** ui5-webcomponents-react Deprecation Notice - ${component} ***\n`;
    // eslint-disable-next-line no-console
    if (console && console.warn) {
      // eslint-disable-next-line no-console
      console.warn(`${value}${message}`);
    }
  }
};

export const enrichEventWithDetails = <T extends Record<string, unknown>, ReturnType = CustomEvent<T>>(
  event: UIEvent,
  payload: T | null = null
) => {
  if (event.hasOwnProperty('persist')) {
    // if there is a persist method, it's an SyntheticEvent so we need to persist it
    event.persist();
  }

  const shouldCreateNewDetails =
    event.detail === null || event.detail === undefined || typeof event.detail !== 'object';
  Object.defineProperty(event, 'detail', {
    value: shouldCreateNewDetails ? {} : event.detail,
    writable: true,
    configurable: true
  });
  Object.assign(event.detail, payload);
  return event as unknown as ReturnType;
};

export { debounce } from './debounce';
