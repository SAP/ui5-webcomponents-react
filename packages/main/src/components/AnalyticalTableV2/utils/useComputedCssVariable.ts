import type { RefObject } from 'react';
import { useEffect, useState } from 'react';

export function useComputedCssVariable<T extends HTMLElement>(ref: RefObject<T>, variableName: string): number | null {
  const [value, setValue] = useState<number | null>(null);

  useEffect(() => {
    if (ref.current) {
      const computedStyles = getComputedStyle(ref.current);
      const variableValue = computedStyles.getPropertyValue(variableName).trim();

      // Attempt to parse the value as a number
      const parsedValue = parseFloat(variableValue);
      setValue(isNaN(parsedValue) ? null : parsedValue);
    }
  }, [variableName, ref]);

  return value;
}
