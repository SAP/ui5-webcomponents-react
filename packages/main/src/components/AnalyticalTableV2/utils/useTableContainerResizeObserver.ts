import type { RefObject } from 'react';
import { useState, useEffect } from 'react';

export const useTableContainerResizeObserver = (tableContainerRef: RefObject<HTMLDivElement>) => {
  const [tableWidth, setTableWidth] = useState(0);
  const [horizontalScrollbarHeight, setHorizontalScrollbarHeight] = useState(0);

  useEffect(() => {
    const tableContainer = tableContainerRef.current;

    const resizeObserver = new ResizeObserver((entries) => {
      for (const entry of entries) {
        if (entry.borderBoxSize) {
          const { borderBoxSize, contentBoxSize } = entry;
          const borderBoxHeight = borderBoxSize[0].blockSize;
          const contentBoxHeight = contentBoxSize[0].blockSize;
          if (borderBoxHeight > contentBoxHeight) {
            setHorizontalScrollbarHeight(borderBoxHeight - contentBoxHeight);
          } else {
            setHorizontalScrollbarHeight(0);
          }

          const borderBoxWidth = borderBoxSize[0].inlineSize;
          setTableWidth(borderBoxWidth);
        }
      }
    });

    if (tableContainer) {
      resizeObserver.observe(tableContainer, { box: 'border-box' });
    }

    return () => {
      if (tableContainer) {
        resizeObserver.unobserve(tableContainer);
      }
    };
  }, [tableContainerRef]);

  return { tableWidth, horizontalScrollbarHeight };
};
