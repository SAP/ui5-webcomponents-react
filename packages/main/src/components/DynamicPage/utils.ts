export const DynamicPageCssVariables = {
  headerDisplay: '--_ui5wcr_DynamicPage_header_display',
  titleFontSize: '--_ui5wcr_DynamicPage_title_fontsize'
};
