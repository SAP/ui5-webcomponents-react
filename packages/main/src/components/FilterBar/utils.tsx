import { cloneElement } from 'react';

const inputTagNames = new Set([
  'UI5-COMBOBOX',
  'UI5-DATE-PICKER',
  'UI5-DATERANGE-PICKER',
  'UI5-DATETIME-PICKER',
  'UI5-DURATION-PICKER',
  'UI5-INPUT',
  'UI5-MULTI-INPUT',
  'UI5-RATING-INDICATOR',
  'UI5-STEP-INPUT',
  'UI5-TIME-PICKER'
]);

export const filterValue = (ref, child) => {
  const tagName = ref.tagName;
  let filterItemProps = {};
  if (inputTagNames.has(tagName)) {
    filterItemProps = { value: ref.value };
  }
  if (tagName === 'UI5-SELECT' || tagName === 'UI5-MULTI-COMBOBOX') {
    const selectedIndices = Array.from(ref.children as HTMLCollectionOf<any>)
      .map((item, index) => (item.selected ? index : false))
      .filter((el) => el !== false);
    const selectedIndicesSet = new Set(selectedIndices);
    const options = child.props.children.props.children?.map((item, index) => {
      if (selectedIndicesSet.has(index)) {
        return cloneElement(item, { selected: true });
      }
      return cloneElement(item, { selected: false });
    });
    filterItemProps = { children: options };
  }
  if (tagName === 'UI5-SWITCH' || tagName === 'UI5-CHECKBOX') {
    filterItemProps = { checked: ref.checked };
  }
  return filterItemProps;
};

export const renderSearchWithValue = (renderSearchElement, searchValue) => {
  return cloneElement(renderSearchElement, { value: searchValue ?? renderSearchElement?.props?.value });
};
