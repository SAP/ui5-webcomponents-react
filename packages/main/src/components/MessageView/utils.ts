import iconAlert from '@ui5/webcomponents-icons/dist/alert.js';
import iconError from '@ui5/webcomponents-icons/dist/error.js';
import iconInformation from '@ui5/webcomponents-icons/dist/information.js';
import iconSysEnter from '@ui5/webcomponents-icons/dist/sys-enter-2.js';
import { ValueState } from '../../enums/ValueState';

export const getIconNameForType = (type: ValueState | keyof typeof ValueState): string => {
  switch (type) {
    case ValueState.Error:
    case 'Error':
      return iconError;
    case ValueState.Success:
    case 'Success':
      return iconSysEnter;
    case ValueState.Warning:
    case 'Warning':
      return iconAlert;
    default:
      return iconInformation;
  }
};
