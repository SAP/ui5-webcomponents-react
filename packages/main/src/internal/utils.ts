import { Children, Fragment, ReactNode } from 'react';

export function flattenFragments(children: ReactNode | ReactNode[], depth = 1): ReactNode[] {
  const flatChildren = [];
  if (Array.isArray(children)) {
    return children;
  }
  const removeFragments = (element, level = 0) => {
    if (!element) {
      return;
    }
    if (element.type === Fragment && level <= depth) {
      Children.toArray(element.props?.children).forEach((item) => {
        removeFragments(item, level + 1);
      });
    } else {
      flatChildren.push(element);
    }
  };
  removeFragments(children);

  return flatChildren;
}
