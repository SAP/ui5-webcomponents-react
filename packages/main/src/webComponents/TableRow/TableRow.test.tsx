import { render } from '@shared/tests';
import { createCustomPropsTest, createWaitForDefineTest } from '@shared/tests/utils';
import React from 'react';
import { TableRow } from './index';

describe('TableRow', () => {
  test('Basic Test (generated)', () => {
    const { asFragment } = render(<TableRow />);
    expect(asFragment()).toMatchSnapshot();
  });
  createCustomPropsTest(TableRow);
  createWaitForDefineTest(TableRow);
});
