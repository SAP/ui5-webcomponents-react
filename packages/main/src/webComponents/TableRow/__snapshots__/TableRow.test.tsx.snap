// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow Basic Test (generated) 1`] = `
<DocumentFragment>
  <ui5-table-row
    type="Inactive"
    ui5-table-row=""
  />
</DocumentFragment>
`;
